### PR TITLE
feat: Implement support for variants

### PR DESCRIPTION
**Pull request by ivarconr**

### CODE REVIEW

File: src/main/java/no/finn/unleash/DefaultUnleash.java [hunk 1]
```diff
@@ -95,6 +100,51 @@ public boolean isEnabled(final String toggleName, final UnleashContext context ,
         return enabled;
     }
 
+    @Override
+    public Variant getVariant(String toggleName, UnleashContext context) {
+        return getVariant(toggleName, context, "");
+    }
+
+    @Override
+    public Variant getVariant(String toggleName, UnleashContext context, String defaultPayload) {
```
Comment (ivarconr) [inline src/main/java/no/finn/unleash/DefaultUnleash.java]: The user should be allowed to inject a default `Variant`, not just the payload.

File: src/main/java/no/finn/unleash/DefaultUnleash.java [hunk 2]
```diff
@@ -95,6 +100,51 @@ public boolean isEnabled(final String toggleName, final UnleashContext context ,
...
+        final  FeatureToggle featureToggle = toggleRepository.getToggle(toggleName);
+        final boolean isEnabled = isEnabled(toggleName, context, false, featureToggle);
+
+        if(isEnabled) {
+            return selectVariant(featureToggle, context)
+                .map(variantDefinition -> new Variant(variantDefinition.getName(), variantDefinition.getPayload(), isEnabled))
+                .orElseGet(() -> new Variant("default", defaultPayload, isEnabled));
```
Comment (ivarconr) [inline src/main/java/no/finn/unleash/DefaultUnleash.java]: The fallback needs to be stream-lined across all client implementations. 

In the code we should also hard-code a fallback-variant, no need to create  a new one every time. 
`Variant fallback = new Variant("disabled", false)`

File: src/main/java/no/finn/unleash/DefaultUnleash.java [hunk 3]
```diff
@@ -95,6 +100,51 @@ public boolean isEnabled(final String toggleName, final UnleashContext context ,
...
+
+        int num = 0;
+        for (VariantDefinition definition: featureToggle.getVariants()){
+            if(score + num >= definition.getWeight()){
+                return Optional.of(definition);
+            }
+            num += definition.getWeight();
+        }
+
+        //Should not happen
```
Comment (ivarconr) [inline src/main/java/no/finn/unleash/DefaultUnleash.java]: What should not happen?

Comment (stalet) [reply src/main/java/no/finn/unleash/DefaultUnleash.java]: Edge case when we are done with the loop. If the result of the modulus operator is higher than max. This is now treated as return empty as we talked about in the discussion this morning.

File: src/main/java/no/finn/unleash/DefaultUnleash.java (hunk 2 above)
Comment (stalet) [reply src/main/java/no/finn/unleash/DefaultUnleash.java]: The payload default payload may differ. We should return the specified default payload on "disabled" right ?

File: src/main/java/no/finn/unleash/DefaultUnleash.java (hunk 1 above)
Comment (stalet) [reply src/main/java/no/finn/unleash/DefaultUnleash.java]: Fixed.

File: src/main/java/no/finn/unleash/DefaultUnleash.java (hunk 2 above)
Comment (stalet) [reply src/main/java/no/finn/unleash/DefaultUnleash.java]: Fixed.

**Comment by sveisvei** [comment]
💯

**Comment by ivarconr** [comment]
only merged to branch so that i can add the override support

**Pull request merged**